ncm-opennebula/src/main/perl/OpenNebula/AII.pm

```diff
@@ -0,0 +1,433 @@
+        $self->verbose("Found RPC section: $section");
+        my $pattern = $config->{$section}->{pattern};
+        if ($pattern and $fqdn =~ /^$pattern$/ and $rpc eq 'rpc') {
+            $rpc = $section;
+            $self->info("Match pattern in RPC section: [$rpc]");
+            last;
+        };
+    };
+    if (exists($config->{$domainname}) and $rpc eq 'rpc') {
+        $rpc = $domainname;
+        $self->info ("Detected configfile RPC section: [$rpc]");
+    };
+    $config->{$rpc}->{port} //= $ONE_DEFAULT_PORT;
+    my $port = $config->{$rpc}->{port};
+    my $host = $config->{$rpc}->{host};
+    $config->{$rpc}->{url} //= $ONE_DEFAULT_URL;
+    $config->{$rpc}->{user} //= $ONEADMIN_USER;
+
+    # Keep backwards compatibility
+    if ($host) {
+        $self->warn("RPC old host section detected: $host. ",
+                              "Please use metaconfig to generate a proper OpenNebula aii configuration.",
+                              "ONE aii will replace the RPC url by the assigned host at this point.");
+        $config->{$rpc}->{url} = "http://$host:$port/RPC2";
+    };
+
+    if (! $config->{$rpc}->{password} ) {
+        $self->error("No password set in configfile $AII_OPENNEBULA_CONFIG. Section [$rpc]");
+        return;
+    };
+
+    my $one = $self->make_one($config->{$rpc});
+    return $one;
+
+}
+
+=item is_supported_one_version
+
+Detects C<OpenNebula> version.
+Returns false if <OpenNebula> version is not supported.
+
+=cut
+
+sub is_supported_one_version
+{
+    my ($self, $one) = @_;
+
+    my $oneversion = $one->version();
+
+    if ($oneversion) {
+        $self->info("Detected OpenNebula version: $oneversion");
+    } else {
+        $self->error("OpenNebula RPC endpoint is not reachable.");
+        return;
+    }
+
+    my $res = $oneversion >= $MINIMAL_ONE_VERSION;
+    if ($res) {
+        $self->verbose("Version $oneversion is ok.");
+        return $oneversion;
+    } else {
+        $self->error("Quattor requires OpenNebula v$MINIMAL_ONE_VERSION or higher (found $oneversion).");
+    }
+    return;
+}
+
+=item get_fqdn
+
+Returns C<fqdn> of the VM
+
+=cut
+
+sub get_fqdn
+{
+    my ($self,$config) = @_;
+    my $hostname = $config->getElement ($HOSTNAME)->getValue;
+    my $domainname = $config->getElement ($DOMAINNAME)->getValue;
+    return "${hostname}.${domainname}";
+}
+
+=item get_resource_instance
+
+Returns ONE virtual resource instance from C<RPC>
+
+=cut
+
+sub get_resource_instance
+{
+    my ($self, $one, $resource, $name) = @_;
+    my $method = "get_${resource}s";
+
+    $method = "get_users" if ($resource eq "owner");
+
+    my @existres = $one->$method(qr{^$name$});
+
+    foreach my $t (@existres) {
+        $self->info("Found requested $resource in ONE database: $name");
+        return $t;
+    };
+    $self->error("Not able to find $resource name $name in ONE database");
+    return;
+}
+
+
+=item is_timeout
+
+Check if the service is removed
```

Comment:
it doesn't matter where it is used, it matters what it does. and it (seems to) check if a resource is avaialbel, with timeout